frb_example/pure_dart_multi/rust/c_output_path/c_output_1.h

```diff
@@ -34,4 +34,11 @@ static int64_t dummy_method_to_enforce_bundling(void) {
     dummy_var ^= ((int64_t) (void*) drop_dart_object);
     dummy_var ^= ((int64_t) (void*) new_dart_opaque);
     return dummy_var;
+}
+#include "c_output_2.h"
+static int64_t dummy_method_to_enforce_bundling(void) {
+    int64_t dummy_var = 0;
+    dummy_var ^= ((int64_t) (void*) dummy_method_to_enforce_bundling_ApiClass1);
+    dummy_var ^= ((int64_t) (void*) dummy_method_to_enforce_bundling_ApiClass2);
+    return dummy_var;
```

Comment:
I am not an expert in C, so the question is: c_output_2.h has dummy_method_to_enforce_bundling_ApiClass2 in *header* file with *static*, and here we are using its *function pointer* instead of actually calling it. So,

1. is it sufficient to forbid the function be stripped? (I guess yes?)
2. is it safe or dangerous to do this? (I guess safe?)

Reply:
I got similar wonders like you, that is why the CI test is essential.

Reply:
> I am not an expert in C, so the question is: c_output_2.h has dummy_method_to_enforce_bundling_ApiClass2 in _header_ file with _static_, and here we are using its _function pointer_ instead of actually calling it. So,
> 
> 1. is it sufficient to forbid the function be stripped? (I guess yes?)
> 2. is it safe or dangerous to do this? (I guess safe?)

I do think this concern is the most important one for this pr.  Frankly, I am not sure what `dummy_method_to_enforce_bundling` acts here.  Intuitionally, I treat it as a "bridge" that tells dart and rust how to translate information between each other.  

And since https://github.com/fzyzcjy/flutter_rust_bridge/releases/tag/v1.59.0, I notice the c headers are only used for compilation within in MacOs or Ios, and it is not used once generation files have been produced by frb in other platforms. So I guess this concern is really something related to MacOs and Ios.  

In addition, I wonder why no issue is posted for multi-blocks for MacOs or Ios, since https://github.com/fzyzcjy/flutter_rust_bridge/releases/tag/v1.59.0 until https://github.com/fzyzcjy/flutter_rust_bridge/issues/1014.

Reply:
Actually, I am thinking of something similar when designing for the shared task. I am wondering if  it is essential and feasible to define an auto-generated header file `shared_auto_c_header.h` to contains only  
```c
#include "c_output_1.h"
#include "c_output_2.h"
static int64_t dummy_method_to_enforce_bundling(void) {
    int64_t dummy_var = 0;
    dummy_var ^= ((int64_t) (void*) dummy_method_to_enforce_bundling_ApiClass1);
    dummy_var ^= ((int64_t) (void*) dummy_method_to_enforce_bundling_ApiClass2);
    return dummy_var;
...
```
for a multi-blocks

Reply:
> I do think this concern is the most important one for this pr. 

Yes - as I mentioned most are nits ;)

> Frankly, I am not sure what dummy_method_to_enforce_bundling acts here.

Shortly speaking, it avoids "symbol stripping" in ios. When ios thinks a function is not used at all, it may throw it away - then Flutter can no longer use that C function. By creating such a weird dummy_method_to_enforce_bundling, we make fake use of those generated C functions, then ios will not strip them.

> So I guess this concern is really something related to MacOs and Ios.

Yes, symbol stripping seems only a problem on ios/macos (if using static lib IIRC)

> In addition, I wonder why no issue is posted for multi-blocks for MacOs or Ios, since [v1.59.0 (release)](https://github.com/fzyzcjy/flutter_rust_bridge/releases/tag/v1.59.0) until https://github.com/fzyzcjy/flutter_rust_bridge/issues/1014.

I guess maybe people do not use this feature that often?

Reply:
stripping: https://docs.flutter.dev/development/platform-integration/ios/c-interop#stripping-ios-symbols , https://cjycode.com/flutter_rust_bridge/integrate/ios_headers.html etc

Reply:
> shared_auto_c_header.h to contains only

Then that may be a breaking change

Reply:
After a second thought, I answer is:
> is it sufficient to forbid the function be stripped? (I guess yes?)

Yes. I think the "strip" issue should be thought of as something related to optimization for c/c++ code related to a specific c/c++ compiler in use. In this way, as long as there is no optimization in the single-block case, the only concern left is the `static` mark for a function symbol.  

As far as I know, when a c/cpp function symbol is marked as `static` excluding the case in a class/struct, it  only means 2 things:
1, there is only 1 instance of such a thing when using it;
2, it is only visible in the current file which defined the static stuff.

That means, with static, there could be functions with the same names in each respective visible scope, like the no-conflict "Parallel universes"([Reference](https://www.scaler.com/topics/static-function-in-c/)). But this is not the case when using FFI here---there is a conflict.

It seems that the `static` has nothing to do with code optimization since the relative address in memory of a function would not be changed whether it is defined as `static` or not. But anyway, I suggest not using `static` in any case(single/multi-blocks), unless it is necessary.

Here, for the multi-block example, I think that marking `dummy_method_to_enforce_bundling_*(void)` without `static` can make them more like a manually defined function API for easier analysis, even for the single-block case.

>is it safe or dangerous to do this? (I guess safe?)

In the aspect of memory, if you are concerning the "dangling pointer", I think it is safe. Because the heap addresses used to initialize/store those API functions would not be changed until APP termination, as long as these addresses are recognized(not optimized out by the c/cpp compiler) to initiate.

Reply:
> marking dummy_method_to_enforce_bundling_*(void) without static

Then if users include it in multi places I am worried maybe it will cause problems? Not an expert in C, but IIRC a function definition in header files should be static b/c this

Reply:
> Then if users include it in multi places I am worried maybe it will cause problems? Not an expert in C, but IIRC a function definition in header files should be static b/c this

I guess [this](https://stackoverflow.com/questions/10847176/should-i-define-static-inline-methods-in-header-file) is the answer.  Roughly, both using `inline` and `static` could avoid conflicts, but the first one is doing so by linking and the latter one is doing so by copying.

So I think it is appropriate to add `inline` for any external linkable function when the function is both declare and defined together in the same file. 

Do I miss something else?